R/project_as_corrected_ppi.R

```diff
@@ -0,0 +1,203 @@
+  scan
+}
+
+
+#' calculate a range-bias corrected PPI
+#'
+#' calculates a PPI that corrects for range-bias effects due to partial beam overlap with
+#' the layer of migration (overshooting) at larger distances from the radar
+#' @inheritParams scan_to_raster
+#' @inheritParams beam_profile_overlap
+#' @param pvol a polar volume of class pvol
+#' @param vp a vertical profile of class vp
+#' @param quantity one or multiple of 'vir','vid','correction_factor', 'overlap'
+#' @param param reflectivity factor scan parameter on which to base range bias corrections.
+#' Typically the same parameter from which animal densities are estimated for object \code{vp}.
+#' One of 'DBZH','DBZV','DBZ','TH','TV'.
+#' @param lat Geodetic latitude of the radar in degrees. If missing taken from \code{pvol}.
+#' @param lon Geodetic latitude of the radar in degrees. If missing taken from \code{pvol}.
+#' @return An object of class '\link[=summary.ppi]{ppi}'.
+#'
+#' @export
+#'
+#' @details to be written
+#' If one of \code{lat} or \code{lon} is missing, the extent of the PPI is take equal to
+#' the extent of the data in the first scan of the polar volume.
+#'
+#' Overlap between vertical profile and vertical radiation profile is calculated using \link{beam_profile}
+#' and stored as quantity 'overlap'.
+#'
+#' scans at 90 degree beam elevation (birdbath scans) are ignored.
+#'
+#' TODO: quantity and param is used in a confusing way, correct.
+#'
+#' @examples
+#' # locate example polar volume file:
+#' pvolfile <- system.file("extdata", "volume.h5", package = "bioRad")
+#' # load polar volume
+#' example_pvol <- read_pvolfile(pvolfile)
+#' # load the corresponding vertical profile for this polar volume
+#' data(example_vp)
+#' # calculate the range-bias corrected ppi on a 100x100 pixel raster
+#' my_ppi <- project_as_corrected_ppi(example_pvol,example_vp,nx=100,ny=100)
+#' # plot the vertically integrated reflectivity (vir) using a 0-2000 cm^2/km^2 color scale:
+#' plot(my_ppi,zlim=c(0,2000))
+#' # calculate the range-bias corrected ppi on finer 1000m x 1000m pixel raster:
+#' my_ppi <- project_as_corrected_ppi(example_pvol,example_vp,res=1000)
+#' # plot the vertically integrated density (vid) using a 0-200 birds/km^2 color scale:
+#' plot(my_ppi,param="vid",zlim=c(0,200))
+#' #' # download a basemap, and map the ppi:
+#' bm = download_basemap(my_ppi)
+#' map(my_ppi, bm)
+#' # calculate the range-bias corrected ppi on an even finer 500m x 500m pixel raster,
+#' # cropping the area up to 50000 meter from the radar.
+#' my_ppi <- project_as_corrected_ppi(example_pvol,example_vp,res=500,
+#' xlim=c(-50000,50000),ylim=c(-50000,50000))
+#' plot(my_ppi,param="vid",zlim=c(0,200))
+project_as_corrected_ppi = function(pvol,vp,nx=100,ny=100,xlim,ylim, zlim=c(0,4000), res, param="DBZH", lat, lon, antenna, beam_angle=1,crs,quantity=c("vir","vid","correction_factor","overlap"), k=4/3, re = 6378, rp = 6357){
```

Comment:
@adokter `project_as_corrected_ppi` and `project_as_ppi` sound so similar to me that I though this could be simplified by adding a "as corrected" parameter to `project_as_ppi`. But it turns out `project_as_corrected_ppi` uses a `pvol` as data input, while `project_as_ppi` only works for `scan` and `param`. Not really sure what to suggest to improve this, but the names are a bit confusing.

Reply:
@peterdesmet I see your point. The output is still a ppi object, which is why I used the `project_as_..._ppi`  name, but definitely a lot more is happening in `project_as_corrected_ppi` than in `project_as_ppi` - it is effectively a vertical integration over all altitudes while accounting for the varying overlap of the radar beam with distance from the radar, not a simple projection.

How about `calculate_vp_corrected_ppi`?

Reply:
Thanks for explaining! The `calculate_vp_etc` name is probably too similar to the very different function `calculate_vp`. I suggest `integrate_to_ppi`:

- Clear that it is more than just projection
- Clear that output is ppi
- Clear that providing a single scan won't do: whole pvol is needed
- Clear that `project_to_ppi` is more basic (and just used for single scans)
- Has similarities to `integrate_profile` which also height integrates

Reply:
I like it, let's go with that!